pkg/ccl/streamingccl/streamingest/stream_ingestion_dist.go

```diff
@@ -0,0 +1,266 @@
+	// Start from the last checkpoint if it exists.
+	if h := progress.GetHighWater(); h != nil && !h.IsEmpty() {
+		previousHighWater = *h
+		heartbeatTimestamp = previousHighWater
+	} else {
+		heartbeatTimestamp = initialScanTimestamp
+	}
+
+	log.Infof(ctx, "ingestion job %d resumes stream ingestion from start time %s",
+		ingestionJob.ID(), heartbeatTimestamp)
+
+	streamID := streampb.StreamID(details.StreamID)
+	updateRunningStatus(ctx, execCtx, ingestionJob, jobspb.InitializingReplication,
+		fmt.Sprintf("connecting to the producer job %d and resuming a stream replication plan", streamID))
+	if err := waitUntilProducerActive(ctx, client, streamID, heartbeatTimestamp, ingestionJob.ID()); err != nil {
+		return err
+	}
+
+	log.Infof(ctx, "producer job %d is active, creating a stream replication plan", streamID)
+	dsp := execCtx.DistSQLPlanner()
+
+	p, planCtx, err := makePlan(
+		execCtx,
+		ingestionJob,
+		details,
+		client,
+		previousHighWater,
+		progress.Details.(*jobspb.Progress_StreamIngest).StreamIngest.Checkpoint,
+		initialScanTimestamp)(ctx, dsp)
+	if err != nil {
+		return err
+	}
+	log.Infof(ctx, "starting to run DistSQL flow for stream ingestion job %d",
+		ingestionJob.ID())
+
+	execPlan := func(ctx context.Context) error {
+		ctx = logtags.AddTag(ctx, "stream-ingest-distsql", nil)
+
+		rw := sql.NewRowResultWriter(nil /* rowContainer */)
+
+		var noTxn *kv.Txn
+		recv := sql.MakeDistSQLReceiver(
+			ctx,
+			rw,
+			tree.Rows,
+			nil, /* rangeCache */
+			noTxn,
+			nil, /* clockUpdater */
+			execCtx.ExtendedEvalContext().Tracing,
+		)
+		defer recv.Release()
+
+		jobsprofiler.StorePlanDiagram(ctx, execCtx.ExecCfg().DistSQLSrv.Stopper, p, execCtx.ExecCfg().InternalDB,
+			ingestionJob.ID())
+
+		// Copy the evalCtx, as dsp.Run() might change it.
+		evalCtxCopy := *execCtx.ExtendedEvalContext()
+		dsp.Run(ctx, planCtx, noTxn, p, recv, &evalCtxCopy, nil /* finishedSetupFn */)
+		return rw.Err()
+	}
+
+	updateRunningStatus(ctx, execCtx, ingestionJob, jobspb.Replicating,
+		"running the SQL flow for the stream ingestion job")
+
+	// TODO(msbutler): implement automatic replanning in the spirit of changefeed replanning.
+	return execPlan(ctx)
+}
+
+// TODO (msbutler): this function signature was written to use in automatic job replanning via
+// sql.PhysicalPlanChangeChecker(). Actually implement c2c replanning.
+func makePlan(
+	execCtx sql.JobExecContext,
+	ingestionJob *jobs.Job,
+	details jobspb.StreamIngestionDetails,
+	client streamclient.Client,
+	previousHighWater hlc.Timestamp,
+	checkpoint jobspb.StreamIngestionCheckpoint,
+	initialScanTimestamp hlc.Timestamp,
+) func(context.Context, *sql.DistSQLPlanner) (*sql.PhysicalPlan, *sql.PlanningCtx, error) {
+	jobID := ingestionJob.ID()
+	return func(ctx context.Context, dsp *sql.DistSQLPlanner) (*sql.PhysicalPlan, *sql.PlanningCtx, error) {
+		log.Infof(ctx, "Re Planning DistSQL flow for stream ingestion job %d", jobID)
+
+		streamID := streampb.StreamID(details.StreamID)
```

Comment:
It doesn't matter, but if you wanted, this and gatewayID below could live outside the returned function along with jobID. Or jobID could live in here.